design-proposals/klip-5-internal-schemas-for-KSQL-engine.md

```diff
@@ -0,0 +1,191 @@
+ the external schema that user content conforms to. The internal KSQL schema will be fully compliant
+ with KSQL requirements and the engine can use it to process queries with no issues. On the other
+ hand, external schema won’t be restricted with KSQL requirements and users can use any content with
+ any naming convention in KSQL.
+
+Users will be able to use quoted identifiers to declare and access fields that do not conform to
+the KSQL identifier requirements. KSQL will generate an internal schema with a one to one mapping
+from the external schema fields into the internal schema fields. All of the computation will use the
+field names from the internal schema.
+
+As an example, consider we want to declare s stream on a topic with values in JSON format. The following
+is a sample message value in this topic:
+
+```json
+{"@ID": 0, "@NAME": "foo", "@AMOUNT": 0.0}
+```
+
+As you can see, two of the field names start with “@” and the third one includes “.” in the field name.
+ Currently we cannot declare a stream or table on this topic since its content does not conform to KQSL
+ field naming requirements. However, we should be able to use quoted identifiers, which is supported by
+ our parser, to declare a stream or table on this topic and write queries over the declared stream or
+ table. Let’s assume we declare a stream on this topic, named foo. The following would be the DDL
+  statement:
+
+```sql
+CREATE STREAM foo
+  (“@ID” BIGINT, “@NAME” STRING, “MESSAGE.AMOUNT” DOUBLE)
+WITH
+  (KAFKA_TOPIC=’foo’, VALUE_FORMAT = ‘JSON’);
+```
+
+KSQL will add the above stream with the declared schema to its metastore. Now, let’s assume we want
+to create a new stream from foo using the following CSAS statement:
+
+```sql
+CREATE STREAM bar AS
+SELECT “@ID”, lowercase(“@NAME”) AS “@NAME”, “MESSAGE.AMOUNT” * 100 AS “AMOUNT.BY.100”
+FROM foo
+WHERE “MESSAGE.AMOUNT” < 1;
+```
+
+KSQL will generate an internal schema for foo before processing the above query and rewrites the query
+based on the internal schema. It then will use the rewritten query to process the messages from the topic.
+
+## Scope
+
+* Support for Quoted Identifiers in DDL and DML statements
+* Creation of internal schema for external schemas
+* Query rewrite to use the internal schemas
+* Query processing using the internal schemas
+
+## Value/Return
+
+This will eliminate the current requirement of having message field names in Kafka topics to conform
+to KSQL internal requirements. With this KLIP, KSQL will be able to process topics with any naming convention.
+
+
+## Public API
+
+The only user facing part in this KLIP is the support for quoted identifier which will be the same
+as the standard quoted identifiers in SQL. The declaration and use of the internal schemas are not
+visible to users and are part of the KSQL engine internals.
+
+## Design
+
+The internal schema will only be used in executing DML statements. When the engine receives a DML
```

Comment:
> This will be done by extending the current KSQLSchema class by adding the internal schema to it.

I recommend we do this by refactoring `LogicalSchema` to no longer use Connect's `Schema` type internally. (I've been doing work to move us in this direction anyway).  

Our own Field class could:
- know if the field is quoted or not, without _encoding_ the quotes into a string).
- Handle field name comparison, taking into account case-sensitivity of quoted identifiers and the SQL standard for them. (It's sufficient to ensure unquoted field names are upper cased and then treat field names as case sensitive).
- Handle full-qualified fields without _encoding_ the source into the name.
- Track any user defined alias. Meaning deserializers can be passed it to map from source to aliased field names.
- Track the new 'internal' field name.
- Use the KSQL type system, rather than Connects, i.e. make use of `SqlType` enum and something like `io.confluent.ksql.parser.tree.Type`s subclasses.

This has many benefits over trying to bolt this onto/into the Connect schema.

Reply:
Totally agree, you have been making lot's of changes on handling schemas and fields and the implementation of this KLIP will use the latest changes. On having our own `Field` class too I agree it would make the implementation of this KLIP much easier.

Reply:
Probably worth mentioning this in the KLIP...